sklearn/neighbors/base.py

```diff
@@ -772,15 +772,24 @@ def fit(self, X, y):
             self.outputs_2d_ = True
 
         check_classification_targets(y)
+
         self.classes_ = []
-        self._y = np.empty(y.shape, dtype=np.int)
-        for k in range(self._y.shape[1]):
-            classes, self._y[:, k] = np.unique(y[:, k], return_inverse=True)
-            self.classes_.append(classes)
-
-        if not self.outputs_2d_:
-            self.classes_ = self.classes_[0]
-            self._y = self._y.ravel()
+
+        if issparse(y) and self.outputs_2d_:
+            self.outputs_2d_ = 'sparse'
+            self._y = y
+            self.classes_ = [np.array([0, 1], dtype=np.int)] * y.shape[1]
```

Comment:
I just mean that, even if (for no particularly good reason) we only support y being sparse when it is multilabel (not general multioutput) someone might get a surprise by this assumption that it's all binary.

You can't just do set(y). if y is CSC or CSR, checking that `np.all(np.union1d(y.data, [0, 1]) == [0, 1])` and raising an error if it fails might help someone unsuspecting.